README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rprimer <img src='man/figures/rprimer.png' align="right" height="139" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/sofpn/rprimer/workflows/R-CMD-check/badge.svg)](https://github.com/sofpn/rprimer/actions)
<!-- badges: end -->

rprimer is an R package that designs degenerate oligos and PCR assays
from a multiple DNA sequence alignment of target sequences of interest.
The package is specifically designed for sequence variable viruses.

## Installation

To install rprimer from
[Bioconductor](https://bioconductor.org/packages/devel/bioc/html/rprimer.html),
start R (version 4.2) and enter:

``` r
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("rprimer")
```

Attach the package by calling:

``` r
library(rprimer)
```

## Overview

The package contains five main functions:

-   `consensusProfile()`
-   `designOligos()`
-   `designAssays()`
-   `checkMatch()`
-   `plotData()`

## Shiny application

The package can be run through a Shiny application (a graphical user
interface). To start the application, type `runRprimerApp()` from within
R upon installing and attaching the package.

The application can also be found online,
[here](https://sofpn.shinyapps.io/rprimer).

## Workflow

### Import alignment

The first step is to import an alignment with target sequences of
interest. This is done by using `readDNAMultipleAlignment()`.

The file “example\_alignment.txt” contains an alignment of 50 hepatitis
E virus sequences.

``` r
infile <- system.file("extdata", "example_alignment.txt", package = "rprimer")

myAlignment <- readDNAMultipleAlignment(infile, format = "fasta")
```

### Step 1: `consensusProfile`

`consensusProfile()` takes a `DNAMultipleAlignment` as input and returns
all the information needed for the subsequent design process.

``` r
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
```

Results (row 100-106):

| position |    a |    c |    g |    t | other | gaps | majority | identity | iupac | coverage |
|---------:|-----:|-----:|-----:|-----:|------:|-----:|:---------|---------:|:------|---------:|
|      100 | 0.00 | 1.00 | 0.00 | 0.00 |  0.00 |    0 | C        |     1.00 | C     |     1.00 |
|      101 | 1.00 | 0.00 | 0.00 | 0.00 |  0.00 |    0 | A        |     1.00 | A     |     1.00 |
|      102 | 0.16 | 0.00 | 0.84 | 0.00 |  0.00 |    0 | G        |     0.84 | R     |     1.00 |
|      103 | 0.00 | 0.00 | 1.00 | 0.00 |  0.00 |    0 | G        |     1.00 | G     |     1.00 |
|      104 | 0.00 | 0.98 | 0.00 | 0.00 |  0.02 |    0 | C        |     1.00 | C     |     1.00 |
|      105 | 0.20 | 0.00 | 0.02 | 0.78 |  0.00 |    0 | T        |     0.78 | W     |     0.98 |
|      106 | 0.00 | 0.00 | 1.00 | 0.00 |  0.00 |    0 | G        |     1.00 | G     |     1.00 |

The results can be visualized with `plotData()`:

``` r
plotData(myConsensusProfile)
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

### Step 2: `designOligos`

The next step is to design oligos. You can either use the default
settings as below, or adjust them as preferred (see the package vignette
or `?designOligos` for more information). The default settings allow a
maximum degeneracy of four, which means that only the most conserved
regions of the genome will be considered as oligo binding sites.

``` r
myOligos <- designOligos(myConsensusProfile)
```

Results (first six rows):

| type   | fwd   | rev   | start | end | length | iupacSequence        | iupacSequenceRc      | identity | coverage | degeneracy | gcContentMean | gcContentRange | tmMean | tmRange | deltaGMean | deltaGRange | sequence   | sequenceRc | gcContent  | tm         | deltaG     | method    | score | roiStart | roiEnd |
|:-------|:------|:------|------:|----:|-------:|:---------------------|:---------------------|---------:|---------:|-----------:|--------------:|---------------:|-------:|--------:|-----------:|------------:|:-----------|:-----------|:-----------|:-----------|:-----------|:----------|------:|---------:|-------:|
| probe  | TRUE  | TRUE  |   124 | 143 |     20 | TCYGCCYTGGCGAATGCTGT | ACAGCATTCGCCARGGCRGA |     0.95 |     0.99 |          4 |          0.60 |           0.10 |  63.17 |    4.33 |     -21.61 |        2.08 | TCCGCCCT…. | ACAGCATT…. | 0.65, 0….. | 65.33623…. | -22.6538…. | ambiguous |     2 |        1 |   7597 |
| probe  | FALSE | TRUE  |   127 | 146 |     20 | GCCYTGGCGAATGCTGTGGT | ACCACAGCATTCGCCARGGC |     0.98 |     0.99 |          2 |          0.62 |           0.05 |  63.18 |    1.84 |     -21.73 |        0.83 | GCCCTGGC…. | ACCACAGC…. | 0.65, 0.6  | 64.10586…. | -22.1475…. | ambiguous |     3 |        1 |   7597 |
| primer | TRUE  | FALSE |   128 | 146 |     19 | CCYTGGCGAATGCTGTGGT  | ACCACAGCATTCGCCARGG  |     0.97 |     0.99 |          2 |          0.61 |           0.05 |  61.48 |    1.95 |     -19.84 |        0.83 | CCCTGGCG…. | ACCACAGC…. | 0.631578…. | 62.45335…. | -20.2570…. | ambiguous |     3 |        1 |   7597 |
| primer | TRUE  | FALSE |   128 | 147 |     20 | CCYTGGCGAATGCTGTGGTR | YACCACAGCATTCGCCARGG |     0.96 |     0.99 |          4 |          0.60 |           0.10 |  61.61 |    3.37 |     -20.55 |        1.76 | CCCTGGCG…. | TACCACAG…. | 0.6, 0.5…. | 61.77166…. | -20.5089…. | ambiguous |     2 |        1 |   7597 |
| probe  | TRUE  | TRUE  |   128 | 146 |     19 | CCYTGGCGAATGCTGTGGT  | ACCACAGCATTCGCCARGG  |     0.97 |     0.99 |          2 |          0.61 |           0.05 |  60.45 |    1.94 |     -19.84 |        0.83 | CCCTGGCG…. | ACCACAGC…. | 0.631578…. | 61.41686…. | -20.2570…. | ambiguous |     3 |        1 |   7597 |
| probe  | TRUE  | TRUE  |   128 | 147 |     20 | CCYTGGCGAATGCTGTGGTR | YACCACAGCATTCGCCARGG |     0.96 |     0.99 |          4 |          0.60 |           0.10 |  60.63 |    3.37 |     -20.55 |        1.76 | CCCTGGCG…. | TACCACAG…. | 0.6, 0.5…. | 60.78676…. | -20.5089…. | ambiguous |     2 |        1 |   7597 |

The results can be visualized as a dashboard, using `plotData()`:

``` r
plotData(myOligos)
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />

### Step 3: `designAssays`

`designAssays()` finds pairs of forward and reverse primers and combine
them with probes, if probes are present in the input dataset. You can
either use the default settings as below, or adjust the design
constraints (see the package vignette or `?designAssays` for more
information).

``` r
myAssays <- designAssays(myOligos)
```

Results (first six rows):

| start |  end | length | totalDegeneracy | score | startFwd | endFwd | lengthFwd | iupacSequenceFwd     | identityFwd | coverageFwd | degeneracyFwd | gcContentMeanFwd | gcContentRangeFwd | tmMeanFwd | tmRangeFwd | deltaGMeanFwd | deltaGRangeFwd | sequenceFwd | gcContentFwd | tmFwd      | deltaGFwd  | methodFwd | startRev | endRev | lengthRev | iupacSequenceRev     | identityRev | coverageRev | degeneracyRev | gcContentMeanRev | gcContentRangeRev | tmMeanRev | tmRangeRev | deltaGMeanRev | deltaGRangeRev | sequenceRev | gcContentRev | tmRev      | deltaGRev  | methodRev | plusPr | minusPr | startPr | endPr | lengthPr | iupacSequencePr        | iupacSequenceRcPr      | identityPr | coveragePr | degeneracyPr | gcContentMeanPr | gcContentRangePr | tmMeanPr | tmRangePr | deltaGMeanPr | deltaGRangePr | sequencePr | sequenceRcPr | gcContentPr | tmPr       | deltaGPr   | methodPr  | roiStart | roiEnd |
|------:|-----:|-------:|----------------:|------:|---------:|-------:|----------:|:---------------------|------------:|------------:|--------------:|-----------------:|------------------:|----------:|-----------:|--------------:|---------------:|:------------|:-------------|:-----------|:-----------|:----------|---------:|-------:|----------:|:---------------------|------------:|------------:|--------------:|-----------------:|------------------:|----------:|-----------:|--------------:|---------------:|:------------|:-------------|:-----------|:-----------|:----------|:-------|:--------|--------:|------:|---------:|:-----------------------|:-----------------------|-----------:|-----------:|-------------:|----------------:|-----------------:|---------:|----------:|-------------:|--------------:|:-----------|:-------------|:------------|:-----------|:-----------|:----------|---------:|-------:|
|  5605 | 5673 |     69 |               6 |  2.00 |     5605 |   5624 |        20 | GGCRGTGGTTTCTGGGGTGA |        0.98 |           1 |             2 |             0.62 |              0.05 |     62.84 |       2.51 |        -20.86 |           1.25 | GGCAGTGG….  | 0.6, 0.65    | 61.57995…. | -20.2350…. | ambiguous |     5654 |   5673 |        20 | GTTGGTTGGATGAASATAGG |           1 |           1 |             2 |              0.4 |                 0 |     50.71 |        1.1 |        -15.27 |           0.52 | GTTGGTTG….  | 0.4, 0.4     | 50.15469…. | -15.0078…. | ambiguous | TRUE   | FALSE   |    5625 |  5642 |       18 | CMGGGTTGATTCTCAGCC     | GGCTGAGAATCAACCCKG     |       0.97 |       0.99 |            2 |            0.58 |             0.06 |    55.14 |      2.79 |       -17.06 |          1.25 | CAGGGTTG…. | GGCTGAGA….   | 0.555555….  | 53.74554…. | -16.4324…. | ambiguous |        1 |   7597 |
|  5605 | 5673 |     69 |               6 |  2.33 |     5605 |   5624 |        20 | GGCRGTGGTTTCTGGGGTGA |        0.98 |           1 |             2 |             0.62 |              0.05 |     62.84 |       2.51 |        -20.86 |           1.25 | GGCAGTGG….  | 0.6, 0.65    | 61.57995…. | -20.2350…. | ambiguous |     5654 |   5673 |        20 | GTTGGTTGGATGAASATAGG |           1 |           1 |             2 |              0.4 |                 0 |     50.71 |        1.1 |        -15.27 |           0.52 | GTTGGTTG….  | 0.4, 0.4     | 50.15469…. | -15.0078…. | ambiguous | TRUE   | FALSE   |    5625 |  5643 |       19 | CMGGGTTGATTCTCAGCCC    | GGGCTGAGAATCAACCCKG    |       0.97 |       0.99 |            2 |            0.61 |             0.05 |    57.63 |      2.64 |       -18.54 |          1.25 | CAGGGTTG…. | GGGCTGAG….   | 0.578947….  | 56.30713…. | -17.9185…. | ambiguous |        1 |   7597 |
|  5605 | 5673 |     69 |               6 |  2.00 |     5605 |   5624 |        20 | GGCRGTGGTTTCTGGGGTGA |        0.98 |           1 |             2 |             0.62 |              0.05 |     62.84 |       2.51 |        -20.86 |           1.25 | GGCAGTGG….  | 0.6, 0.65    | 61.57995…. | -20.2350…. | ambiguous |     5654 |   5673 |        20 | GTTGGTTGGATGAASATAGG |           1 |           1 |             2 |              0.4 |                 0 |     50.71 |        1.1 |        -15.27 |           0.52 | GTTGGTTG….  | 0.4, 0.4     | 50.15469…. | -15.0078…. | ambiguous | TRUE   | TRUE    |    5625 |  5644 |       20 | CMGGGTTGATTCTCAGCCCT   | AGGGCTGAGAATCAACCCKG   |       0.97 |       1.00 |            2 |            0.58 |             0.05 |    58.87 |      2.55 |       -19.43 |          1.25 | CAGGGTTG…. | AGGGCTGA….   | 0.55, 0.6   | 57.59836…. | -18.8035…. | ambiguous |        1 |   7597 |
|  5605 | 5673 |     69 |               6 |  1.67 |     5605 |   5624 |        20 | GGCRGTGGTTTCTGGGGTGA |        0.98 |           1 |             2 |             0.62 |              0.05 |     62.84 |       2.51 |        -20.86 |           1.25 | GGCAGTGG….  | 0.6, 0.65    | 61.57995…. | -20.2350…. | ambiguous |     5654 |   5673 |        20 | GTTGGTTGGATGAASATAGG |           1 |           1 |             2 |              0.4 |                 0 |     50.71 |        1.1 |        -15.27 |           0.52 | GTTGGTTG….  | 0.4, 0.4     | 50.15469…. | -15.0078…. | ambiguous | TRUE   | TRUE    |    5625 |  5645 |       21 | CMGGGTTGATTCTCAGCCCTT  | AAGGGCTGAGAATCAACCCKG  |       0.98 |       1.00 |            2 |            0.55 |             0.05 |    59.21 |      2.43 |       -20.08 |          1.25 | CAGGGTTG…. | AAGGGCTG….   | 0.523809….  | 57.99472…. | -19.4553…. | ambiguous |        1 |   7597 |

NEWS.md
Changes in version 0.99.0 (2021-10-08)  
+ Package was submitted to Bioconductor
Changes in version 0.99.5 (2021-11-12)
+ Package was accepted to Bioconductor
Changes in version 1.5.0
+ The vignette has been updated for clarification
+ The maximum allowed degeneracy for primers and probes has been increased


R/plots.R
#' Plot an RprimerProfile-object (method)
#'
#' @param x An \code{RprimerProfile} object.
#'
#' @param shadeFrom
#' Optional. If a particular area should be shaded, where the shade
#' should begin (an integer).
#'
#' @param shadeTo
#' Optional. If a particular area should be shaded, where the shade
#' should end (an integer).
#'
#' @return A plot.
#'
#' @describeIn plotData
#'
#' @importFrom patchwork wrap_plots
#'
#' @export
setMethod("plotData", "RprimerProfile", function(x,
                                                 shadeFrom = NULL,
                                                 shadeTo = NULL) {
    if (is.null(shadeFrom) && is.null(shadeTo)) {
        shadeFrom <- -Inf
        shadeTo <- -Inf
    }
    if (is.null(shadeFrom) && !is.null(shadeTo)) {
        shadeFrom <- 0
    }
    if (!is.null(shadeFrom) && is.null(shadeTo)) {
        shadeTo <- nrow(x)
    }
    if (!is.null(shadeFrom) && !is.numeric(shadeFrom)) {
        stop("'shadeFrom' must be a number.", call. = FALSE)
    }
    if (!is.null(shadeTo) && !is.numeric(shadeTo)) {
        stop("'shadeTo' must be a number.", call. = FALSE)
    }
    x <- as.data.frame(x)
    patchwork::wrap_plots(
        list(
            .identityPlot(x, shadeFrom = shadeFrom, shadeTo = shadeTo),
            .entropyPlot(x, shadeFrom = shadeFrom, shadeTo = shadeTo),
            .gcPlot(x, shadeFrom = shadeFrom, shadeTo = shadeTo),
            .gapPlot(x, shadeFrom = shadeFrom, shadeTo = shadeTo)
        ),
        ncol = 1
    )
})

#' Plot an RprimerOligo-object (method)
#'
#' @param x An \code{RprimerOligo} object.
#'
#' @return A plot.
#'
#' @describeIn plotData
#'
#' @export
setMethod("plotData", "RprimerOligo", function(x) {
    x <- as.data.frame(x)
    patchwork::wrap_plots(
        list(.oligoPlot(x), .oligoFeaturePlot(x)),
        ncol = 1
    )
})

#' Plot an RprimerAssay-object (method)
#'
#' @param x An \code{RprimerAssay} object.
#'
#' @return A plot.
#'
#' @describeIn plotData
#'
#' @export
setMethod("plotData", "RprimerAssay", function(x) {
    x <- as.data.frame(x)
    patchwork::wrap_plots(
        list(.assayPlot(x), .assayFeaturePlot(x)),
        ncol = 1
    )
})

#' Plot nucleotide distribution
#'
#' @param x An \code{RprimerProfile} object.
#'
#' @param rc
#' \code{TRUE} or {FALSE}. If the plotted sequence should be displayed
#' as reverse complement or not. Defaults to \code{FALSE}.
#'
#' @return A barplot with nucleotide distribution.
#'
#' @examples
#' data("exampleRprimerProfile")
#' plotNucleotides(exampleRprimerProfile[1:10, ])
#' ## Plot the first 10 bases
#' plotNucleotides(exampleRprimerProfile[1:10, ], rc = TRUE)
#' ## As reverse complement
#' @export
plotNucleotides <- function(x, rc = FALSE) {
    if (!methods::is(x, "RprimerProfile")) {
        stop("'x' must be an RprimerProfile object.", call. = FALSE)
    }
    if (!(is.logical(rc))) {
        stop("'rc' must be set to TRUE or FALSE.", call. = FALSE)
    }
    position <- x$position
    x <- as.data.frame(x)
    x <- dplyr::select(x, c("a", "c", "g", "t", "other"))
    x <- t(as.matrix(x))
    colnames(x) <- position
    rownames(x)[-5] <- toupper(rownames(x))[-5]
    if (rc) {
        rownames(x) <- unname(rprimerGlobals$complementLookup[rownames(x)])
        x <- x[order(rownames(x)), ]
        if ("other" %in% rownames(x)) {
            swap <- c(which(rownames(x) == "other"), nrow(x))
            x[swap, ] <- x[rev(swap), ]
            rownames(x)[swap] <- rownames(x)[rev(swap)]
        }
    }
    Base <- Position <- Frequency <- NULL
    x <- reshape2::melt(x)
    names(x) <- c("Base", "Position", "Frequency")
    x$Base <- factor(x$Base)
    x$Position <- factor(x$Position)
    if (rc) {
        x$Position <- factor(x$Position, levels = rev(levels(x$Position)))
    }
    basePalette <- c(
        "other" = "#E8E9ED", "A" = "#FFC759", "C" = "#FF7B9C",
        "G" = "#607196", "T" = "#BABFD1"
    )
    ggplot2::ggplot(
        data = x, ggplot2::aes(x = Position, y = Frequency, fill = Base)
    ) +
        ggplot2::geom_bar(stat = "identity") +
        ggplot2::scale_fill_manual(values = basePalette) +
        .themeRprimer(showLegend = TRUE, rotateX = TRUE)
}

# Helpers ======================================================================

.oligoPlot <- function(x) {
    if (any(x$type == "probe")) {
        .primerProbePlot(x)
    } else {
        .primerPlot(x)
    }
}

.primerPlot <- function(x) {
    start <- x$alignmentStart[[1]]
    end <- x$alignmentEnd[[1]]
    colors <- c(
        fwd = "#A4A7B6", rev = "#82879B", prPos = "#64697D", prNeg = "#525666"
    )
    ggplot2::ggplot() +
        ggplot2::xlim(start, end) +
        ggplot2::ylim(0, 1) +
        ggplot2::labs(x = "Position", y = "") +
        ggplot2::geom_segment(
            color = "grey", lwd = 2,
            ggplot2::aes(x = start, xend = end, y = 0, yend = 0)
        ) +
        ggplot2::geom_rect(
            data = x[x$type == "primer" & !is.na(x$majority), ], ggplot2::aes(
                xmin = start, xmax = end, ymin = 0.35, ymax = 0.65
            ), fill = colors["fwd"]
        ) +
        ggplot2::geom_rect(
            data = x[x$type == "primer" & !is.na(x$majorityRc), ],
            ggplot2::aes(
                xmin = start, xmax = end, ymin = 0.05, ymax = 0.35
            ), fill = colors["rev"]
        ) +
        ggplot2::annotate(
            "label",
            x = start, y = seq(0.89, length.out = 2, by = 0.07),
            label = c(
                paste(
                    "Reverse primer n =",
                    nrow(x[x$type == "primer" & !is.na(x$majorityRc), ])
                ),
                paste(
                    "Forward primer n =",
                    nrow(x[x$type == "primer" & !is.na(x$majority), ])
                )
            ), size = 3, hjust = 0, fontface = 2,
            color = rev(colors[seq_len(2)]), fill = "white", label.size = NA
        ) +
        .themeRprimer(showYAxis = FALSE)
}


.primerProbePlot <- function(x) {
    start <- x$alignmentStart[[1]]
    end <- x$alignmentEnd[[1]]
    colors <- c(
        fwd = "#A4A7B6", rev = "#82879B", prPos = "#64697D", prNeg = "#525666"
    )
    ggplot2::ggplot() +
        ggplot2::xlim(start, end) +
        ggplot2::ylim(0, 1) +
        ggplot2::labs(x = "Position", y = "") +
        ggplot2::geom_segment(
            color = "grey", lwd = 2,
            ggplot2::aes(x = start, xend = end, y = 0, yend = 0)
        ) +
        ggplot2::geom_rect(
            data = x[x$type == "primer" & !is.na(x$majority), ], ggplot2::aes(
                xmin = start, xmax = end, ymin = 0.5, ymax = 0.65
            ), fill = colors["fwd"]
        ) +
        ggplot2::geom_rect(
            data = x[x$type == "primer" & !is.na(x$majorityRc), ],
            ggplot2::aes(
                xmin = start, xmax = end, ymin = 0.35, ymax = 0.5
            ), fill = colors["rev"]
        ) +
        ggplot2::geom_rect(
            data = x[x$type == "probe" & !is.na(x$majority), ], ggplot2::aes(
                xmin = start, xmax = end, ymin = 0.20, ymax = 0.35
            ), fill = colors["prPos"]
        ) +
        ggplot2::geom_rect(
            data = x[x$type == "probe" & !is.na(x$majorityRc), ], ggplot2::aes(
                xmin = start, xmax = end, ymin = 0.05, ymax = 0.20
            ), fill = colors["prNeg"]
        ) +
        ggplot2::annotate(
            "label",
            x = start, y = seq(0.75, length.out = 4, by = 0.07),
            label = c(
                paste(
                    "Probe (-) n =",
                    nrow(x[x$type == "probe" & !is.na(x$majorityRc), ])
                ),
                paste(
                    "Probe (+) n =",
                    nrow(x[x$type == "probe" & !is.na(x$majority), ])
                ),
                paste(
                    "Reverse primer n =",
                    nrow(x[x$type == "primer" & !is.na(x$majorityRc), ])
                ),
                paste(
                    "Forward primer n =",
                    nrow(x[x$type == "primer" & !is.na(x$majority), ])
                )
            ), size = 3, hjust = 0, fontface = 2,
            color = rev(colors), fill = "white", label.size = NA
        ) +
        .themeRprimer(showYAxis = FALSE)
}

.violinPlot <- function(data, y, title = "", color = "grey30") {
    ggplot2::ggplot() +
        ggplot2::geom_violin(
            data = data, ggplot2::aes(x = 1, y = y),
            fill = color, color = NA, alpha = 0.4
        ) +
        ggplot2::geom_point(
            data = data, ggplot2::aes(x = 1, y = y), alpha = 0.05, color = color
        ) +
        ggplot2::geom_boxplot(
            data = data, ggplot2::aes(x = 1, y = y), width = 0.1,
            color = color, fill = color, alpha = 0.2
        ) +
        ggplot2::ylab("") +
        ggplot2::labs(title = title) +
        .themeRprimer(showXAxis = FALSE)
}

.barPlot <- function(data, y, title = "", color = "grey30") {
    ggplot2::ggplot(data, ggplot2::aes(factor(y))) +
        ggplot2::geom_bar(
            fill = color, color = color, alpha = 0.4
        ) +
        ggplot2::xlab("") +
        ggplot2::ylab("") +
        ggplot2::labs(title = title) +
        .themeRprimer(showXAxis = TRUE, rotateX = TRUE)
}

.gcTmIdentityPlot <- function(x, color = "grey30", type = "Primers") {
    patchwork::wrap_plots(
        list(
            .violinPlot(
                x, x$gcMajority,
                paste0("\n", type, "\n\nGC-content"),
                color = color
            ),
            .violinPlot(x, x$tmMajority, "\n\n\nTm", color = color),
            .violinPlot(x, x$identity, "\n\n\nIdentity", color = color),
            .barPlot(x, x$length, "\n\n\nLength", color = color),
            .barPlot(x, x$degeneracy, "\n\n\nDegeneracy", color = color)
        ),
        ncol = 5
    )
}

.oligoFeaturePlot <- function(x) {
    if (any(x$type == "probe")) {
        patchwork::wrap_plots(list(
            .gcTmIdentityPlot(
                x[x$type == "primer", ],
                color = "grey30", type = "Primers"
            ),
            .gcTmIdentityPlot(
                x[x$type == "probe", ],
                color = "grey30", type = "Probes"
            )
        ), ncol = 1)
    } else {
        patchwork::wrap_plots(list(
            .gcTmIdentityPlot(
                x[x$type == "primer", ],
                color = "grey30",
                type = "Primers"
            )
        ), ncol = 1)
    }
}

.assayPlot <- function(x) {
    start <- x$alignmentStart[[1]]
    end <- x$alignmentEnd[[1]]
    row <- seq_len(nrow(x))
    x <- tibble::add_column(x, row)
    ggplot2::ggplot() +
        ggplot2::xlim(start, end) +
        ggplot2::ylim(0, 1) +
        ggplot2::labs(x = "Position", y = "") +
        ggplot2::geom_segment(
            color = "grey", lwd = 2,
            ggplot2::aes(x = start, xend = end, y = 0, yend = 0)
        ) +
        ggplot2::geom_rect(
            data = x, ggplot2::aes(
                xmin = start, xmax = end, ymin = 0.05, ymax = 0.65
            ), fill = "#64697D"
        ) +
        ggplot2::annotate(
            "label",
            x = start, y = 0.8,
            label = paste(
                "Assays n =", nrow(x)
            ), size = 3, hjust = 0, fontface = 2,
            color = "#64697D", fill = "white", label.size = NA
        ) +
        .themeRprimer(showYAxis = FALSE)
}

.assayFeaturePlot <- function(x, color = "grey30") {
    patchwork::wrap_plots(
        list(
            .violinPlot(
                x, x$tmDifferencePrimer, "\n\nTm diff., primers",
                color = color
            ),
            .violinPlot(
                x, x$tmDifferencePrimerProbe, "\n\nTm diff., primers-probe",
                color = color
            ),
            .violinPlot(x, x$meanIdentity, "\n\nMean identity", color = color),
            .barPlot(
                x, x$ampliconLength, "\n\nAmplicon length",
                color = color
            ),
            .barPlot(
                x, x$totalDegeneracy, "\n\nTotal degeneracy",
                color = color
            )
        ),
        ncol = 5
    )
}

.identityPlot <- function(x, shadeFrom = NULL, shadeTo = NULL) {
    position <- identity <- average <- NULL
    averages <- .runningAverage(x$identity)
    xadj <- unique(x$position - seq_along(x$position))
    averages$position <- averages$position + xadj
    ggplot2::ggplot(
        data = x, ggplot2::aes(x = position, y = identity)
    ) +
        .shadeArea(shadeFrom = shadeFrom, shadeTo = shadeTo) +
        ggplot2::geom_point(size = 0.3, ggplot2::aes(colour = identity)) +
        ggplot2::geom_line(
            data = averages,  color = "#1B1C22",
            ggplot2::aes(x = position, y = average)
        ) +
        ggplot2::ylim(0, 1) +
        ggplot2::ylab("Identity") +
        ggplot2::xlab("") +
        ggplot2::scale_color_gradient(low = "#BABFD1", high = "#BABFD1") +
        .themeRprimer(showXAxis = FALSE)
}

.entropyPlot <- function(x, shadeFrom = NULL, shadeTo = NULL) {
    position <- entropy <- average <- NULL
    averages <- .runningAverage(x$entropy)
    xadj <- unique(x$position - seq_along(x$position))
    averages$position <- averages$position + xadj
    ggplot2::ggplot(
        data = x, ggplot2::aes(x = position, y = entropy)
    ) +
        .shadeArea(shadeFrom = shadeFrom, shadeTo = shadeTo) +
        ggplot2::geom_point(size = 0.3, ggplot2::aes(colour = entropy)) +
        ggplot2::geom_line(
            data = averages,  color = "#1B1C22",
            ggplot2::aes(x = position, y = average)
        ) +
        ggplot2::ylab("Entropy") +
        ggplot2::xlab("") +
        ggplot2::scale_color_gradient(low = "#BABFD1", high = "#BABFD1") +
        .themeRprimer(showXAxis = FALSE)
}

.gcPlot <- function(x, shadeFrom = NULL, shadeTo = NULL) {
    position <- average <- NULL
    averages <- .gcRunningAverage(x$majority)
    xadj <- unique(x$position - seq_along(x$position))
    averages$position <- averages$position + xadj
    ggplot2::ggplot(data = x, ggplot2::aes(x = position)) +
        ggplot2::geom_segment(
            color = "#BABFD1",
            ggplot2::aes(
                x = min(position), xend = max(position), y = 0.5, yend = 0.5
            )
        ) +
        .shadeArea(shadeFrom = shadeFrom, shadeTo = shadeTo) +
        ggplot2::geom_line(
            data = averages, color = "#1B1C22",
            ggplot2::aes(x = position, y = average)
        ) +
        ggplot2::xlab("") +
        ggplot2::ylab("GC-content") +
        ggplot2::ylim(0, 1) +
        .themeRprimer(showXAxis = FALSE)
}

.gapPlot <- function(x, shadeFrom = NULL, shadeTo = NULL) {
    position <- gaps <- NULL
    ggplot2::ggplot(
        data = x, ggplot2::aes(x = position, y = gaps)
    ) +
        .shadeArea(shadeFrom = shadeFrom, shadeTo = shadeTo) +
        ggplot2::geom_point(size = 0.3, ggplot2::aes(color = gaps)) +
        ggplot2::ylim(0, 1) +
        ggplot2::xlab("Position") +
        ggplot2::ylab("Gaps") +
        ggplot2::scale_color_gradient(low = "#BABFD1", high = "#BABFD1") +
        .themeRprimer()
}

.shadeArea <- function(shadeFrom = NULL, shadeTo = NULL) {
    ggplot2::annotate(
        "rect",
        xmin = shadeFrom, xmax = shadeTo, ymin = -Inf, ymax = Inf,
        color = "white", alpha = 0.2, fill = "grey30"
    )
}

.themeRprimer <- function(showXAxis = TRUE,
                          showYAxis = TRUE,
                          showLegend = FALSE,
                          rotateX = FALSE) {
    if (showXAxis && showYAxis) {
        .showXYAxes(showLegend = showLegend)
    } else if (showXAxis && !showYAxis) {
        .showXAxisHideYAxis(showLegend = showLegend)
    } else if (!showXAxis && showYAxis) {
        .showYAxisHideXAxis(showLegend = showLegend)
    } else {
        .hideXYAxes(showLegend = showLegend)
    }
}

.showXYAxes <- function(showLegend = TRUE, rotateX = FALSE) {
    ggplot2::theme_light() +
        ggplot2::theme(
            legend.title = ggplot2::element_blank(),
            legend.text = ggplot2::element_text(color = "grey30", size = 9),
            legend.position = ifelse(showLegend, "right", "none"),
            axis.title.y = ggplot2::element_text(
                size = 9, color = "grey30",
                margin = ggplot2::margin(r = 10)
            ),
            axis.title.x = ggplot2::element_text(
                size = 9, color = "grey30",
                margin = ggplot2::margin(t = 10)
            ),
            axis.text.x = ggplot2::element_text(
                angle = ifelse(rotateX, 90, 0), vjust = 0.5, hjust = 1
            ),
            plot.title = ggplot2::element_text(size = 9, color = "grey30"),
            plot.margin = ggplot2::unit(rep(0.1, 4), "cm")
        )
}

.showXAxisHideYAxis <- function(showLegend = TRUE, rotateX = FALSE) {
    ggplot2::theme_light() +
        ggplot2::theme(
            legend.title = ggplot2::element_blank(),
            legend.text = ggplot2::element_text(color = "grey30", size = 9),
            legend.position = ifelse(showLegend, "right", "none"),
            axis.title.x = ggplot2::element_text(
                size = 9, color = "grey30",
                margin = ggplot2::margin(t = 10)
            ),
            axis.text.x = ggplot2::element_text(
                angle = ifelse(rotateX, 90, 0), vjust = 0.5, hjust = 1
            ),
            axis.title.y = ggplot2::element_blank(),
            axis.text.y = ggplot2::element_blank(),
            axis.ticks.y = ggplot2::element_blank(),
            plot.title = ggplot2::element_text(size = 9, color = "grey30"),
            plot.margin = ggplot2::unit(rep(0.1, 4), "cm")
        )
}

.showYAxisHideXAxis <- function(showLegend = TRUE) {
    ggplot2::theme_light() +
        ggplot2::theme(
            legend.title = ggplot2::element_blank(),
            legend.text = ggplot2::element_text(color = "grey30", size = 9),
            legend.position = ifelse(showLegend, "right", "none"),
            axis.title.y = ggplot2::element_text(
                size = 9, color = "grey30",
                margin = ggplot2::margin(r = 10)
            ),
            axis.title.x = ggplot2::element_blank(),
            axis.text.x = ggplot2::element_blank(),
            axis.ticks.x = ggplot2::element_blank(),
            plot.title = ggplot2::element_text(size = 9, color = "grey30"),
            plot.margin = ggplot2::unit(rep(0.1, 4), "cm")
        )
}

.hideXYAxes <- function(showLegend = TRUE) {
    ggplot2::theme_light() +
        ggplot2::theme(
            legend.title = ggplot2::element_blank(),
            legend.text = ggplot2::element_text(color = "grey30", size = 9),
            legend.position = ifelse(showLegend, "right", "none"),
            axis.title.y = ggplot2::element_blank(),
            axis.text.y = ggplot2::element_blank(),
            axis.ticks.y = ggplot2::element_blank(),
            axis.title.x = ggplot2::element_blank(),
            axis.text.x = ggplot2::element_blank(),
            axis.ticks.x = ggplot2::element_blank(),
            plot.title = ggplot2::element_text(size = 9, color = "grey30"),
            plot.margin = ggplot2::unit(rep(0.1, 4), "cm")
        )
}
